iocore/net/QUICNetVConnection.cc

```diff
@@ -1340,125 +1352,56 @@ QUICNetVConnection::_packetize_frames(QUICEncryptionLevel level, uint64_t max_pa
-  if (frame) {
-    ++frame_count;
-    probing |= frame->is_probing_frame();
-    this->_store_frame(buf, len, max_frame_size, frame, frames);
-  }
-
-  // DATA_BLOCKED
-  if (this->_remote_flow_controller->credit() == 0 && this->_stream_manager->will_generate_frame(level)) {
-    frame = this->_remote_flow_controller->generate_frame(level, UINT16_MAX, max_frame_size);
-    if (frame) {
-      ++frame_count;
-      probing |= frame->is_probing_frame();
-      this->_store_frame(buf, len, max_frame_size, frame, frames);
-    }
-  }
-
-  // STREAM, MAX_STREAM_DATA, STREAM_DATA_BLOCKED
-  if (!this->_path_validator->is_validating()) {
-    frame = this->_stream_manager->generate_frame(level, this->_remote_flow_controller->credit(), max_frame_size);
-    while (frame) {
-      ++frame_count;
-      probing |= frame->is_probing_frame();
-      if (frame->type() == QUICFrameType::STREAM) {
-        int ret = this->_remote_flow_controller->update(this->_stream_manager->total_offset_sent());
-        QUICFCDebug("[REMOTE] %" PRIu64 "/%" PRIu64, this->_remote_flow_controller->current_offset(),
-                    this->_remote_flow_controller->current_limit());
-        ink_assert(ret == 0);
+  bool ack_only = true;
+  for (auto g : this->_frame_generators) {
+    while (g->will_generate_frame(level)) {
+      // FIXME will_generate_frame should receive more parameters so we don't need extra checks
+      if (g == this->_remote_flow_controller && !this->_stream_manager->will_generate_frame(level)) {
+        break;
       }
-      this->_store_frame(buf, len, max_frame_size, frame, frames);
-
-      if (++this->_stream_frames_sent % MAX_CONSECUTIVE_STREAMS == 0) {
+      if (g == this->_stream_manager && this->_path_validator->is_validating()) {
         break;
       }
 
-      frame = this->_stream_manager->generate_frame(level, this->_remote_flow_controller->credit(), max_frame_size);
-    }
-  }
+      // Common block
+      frame = g->generate_frame(level, this->_remote_flow_controller->credit(), max_frame_size);
+      if (frame) {
+        ++frame_count;
+        probing |= frame->is_probing_frame();
+        if (frame->is_flow_controlled()) {
+          int ret = this->_remote_flow_controller->update(this->_stream_manager->total_offset_sent());
+          QUICFCDebug("[REMOTE] %" PRIu64 "/%" PRIu64, this->_remote_flow_controller->current_offset(),
+                      this->_remote_flow_controller->current_limit());
+          ink_assert(ret == 0);
+        }
+        this->_store_frame(buf, len, max_frame_size, frame, frames);
 
-  // ACK
-  if (frame_count == 0) {
-    if (this->_ack_frame_manager.will_generate_frame(level)) {
-      frame = this->_ack_frame_manager.generate_frame(level, UINT16_MAX, max_frame_size);
-    }
-  } else {
-    frame = this->_ack_frame_manager.generate_frame(level, UINT16_MAX, max_frame_size);
-  }
+        // FIXME ACK frame should have priority
+        if (frame->type() == QUICFrameType::STREAM) {
```

Comment:
Prior this change, it looks like below block was not under this condition.

Reply:
Probably, this is fine.

Reply:
Yeah, I think it's a bug because MAX_STREAM_DATA and STREAM_DATA_BLOCKED increment the counter too.